### PR TITLE
8333182: Add truncated tracing mode for TraceBytecodes

### PR DESCRIPTION
**Pull request by cl4es**
We routinely use `-XX:+TraceBytecodes` to instrument what's going on during application startup. However, much of the information traced in this mode is not really used, and adding a mode which truncates non-relevant information would be helpful to speed up instrumentation, especially on larger apps.

To illustrate, running `-XX:+TraceBytecodes` on a version of netty can produce more than 844MB of output:
```
14759490 bytecodes executed in 151.1s (0.098MHz)
[BytecodeCounter::counter_value = 14759531]
```

A sample output block looks like this:
```
[34051] static jobject sun.reflect.annotation.AnnotationParser.parseAnnotations2(jobject, jobject, jobject, jobject)
[34051] 14759490    58  astore #9
[34051] 14759491    60  aload #9
[34051] 14759492    62  invokestatic 57 
... 19 more lines
[34051] 14759512    41  invokestatic 47 <sun/reflect/annotation/AnnotationParser.parseAnnotation2(Ljava/nio/ByteBuffer;Ljdk/internal/reflect/ConstantPool;Ljava/lang/Class;Z[Ljava/lang/Class;)Ljava/lang/annotation/Annotation;> 

[34051] static jboolean sun.reflect.annotation.AnnotationParser.contains(jobject, jobject)
[34051] 14759513     0  nofast_aload_0
[34051] 14759514     1  astore_2
[34051] 14759515     2  aload_2
...
```

This PR add a develop flag, `TraceBytecodesTruncated` meant to be used in addition to `TraceBytecodes`. It has the effect that all but the first bytecode in each method transition as well as returns and throws will be omitted. With the patch the output may look like this:

```
[42243] static jobject sun.reflect.annotation.AnnotationParser.parseAnnotations2(jobject, jobject, jobject, jobject)
[42243] 14858715    58  astore

[42243] static jboolean sun.reflect.annotation.AnnotationParser.contains(jobject, jobject)
[42243] 14858737     0  nofast_aload_0
[42243] 14858772    35  ireturn
```

This truncated output is sufficient for certain purposes such as when digesting the output with [bytestacks](https://github.com/cl4es/bytestacks).

Running the same application with the patch and the added flag, `-XX:+TraceBytecodes -XX:+TraceBytecodesTruncated` generates a file that is 177MB, and in about a third of the time:
```
14867664 bytecodes executed in 52.4s (0.284MHz)
[BytecodeCounter::counter_value = 14867704]
```

This enhancement would greatly speed up some of our diagnostics, and enable using these tools on even larger apps.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8333182](https://bugs.openjdk.org/browse/JDK-8333182): Add truncated tracing mode for TraceBytecodes (**Enhancement** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Frederic Parain](https://openjdk.org/census#fparain) (@fparain - **Reviewer**) ⚠️ Review applies to [cab354e9](https://git.openjdk.org/jdk/pull/19457/files/cab354e925b1418adcd6a76e55528960a0b5a991)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [cab354e9](https://git.openjdk.org/jdk/pull/19457/files/cab354e925b1418adcd6a76e55528960a0b5a991)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19457/head:pull/19457` \
`$ git checkout pull/19457`

Update a local copy of the PR: \
`$ git checkout pull/19457` \
`$ git pull https://git.openjdk.org/jdk.git pull/19457/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19457`

View PR using the GUI difftool: \
`$ git pr show -t 19457`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19457.diff">https://git.openjdk.org/jdk/pull/19457.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19457#issuecomment-2137726038)

### CODE REVIEW

**Review by dholmes-ora — COMMENTED** [review]
I was surprised not to find a RFE to convert TraceBytecodes to Unified Logging. That would solve your problem as you could just reclassify the too-verbose "logging" statements. Not to put you on-the-spot @cl4es  but do you have a sense for how difficult a conversion to UL would be? :)

**Comment by cl4es** [comment]
The tracer code isn't all that large or convoluted, but some of the printing code is shared with `Method::print_codes_on`, which is used in a number of places in product code. Some of the use cases of that, like the one in idealGraphPrinter.cpp, doesn't seem easily portable to UL since we're streaming the output into some XML document. Of course we could disentangle all that if we accept a bit of code duplication. 

Porting to UL also adds a bit of (optional) logging to each line, growing the output and requiring parser changes. Perhaps there are some ways to tame UL to output exactly - or almost exactly - the same thing, but regardless it definitely complicates parsing. That also begs the question if we need to retain `-XX:+TraceBytecodes` and the exact format it outputs in case there are other consumers out there. I don't know. 

I don't have a strong opinion. I do _really_ want this feature in ASAP (to speed up my work and make the startup tracing useful on larger apps), but I have no idea how high of a priority it is to port this to UL.

**Review by dholmes-ora — APPROVED** [review]
Okay it sounds like UL is not a feasible short term solution here., so being able to truncate the verbose output seems reasonable.

**Comment by cl4es** [comment]
Thanks!

**Comment by cl4es** [comment]
Any second reviewer for this? Or is it trivial enough to go ahead with one?

**Comment by coleenp** [comment]
Its not really trivial.  I'll have a look today.

**Review by fparain — APPROVED** [review]
LGTM.

File: src/hotspot/share/interpreter/bytecodeTracer.cpp [hunk 1]
```diff
@@ -96,7 +96,8 @@ class BytecodePrinter {
   // the adjustments that BytecodeStream performs applies.
   void trace(const methodHandle& method, address bcp, uintptr_t tos, uintptr_t tos2, outputStream* st) {
     ResourceMark rm;
-    if (_current_method != method()) {
+    bool methodChange = _current_method != method();
```
Comment (coleenp) [inline src/hotspot/share/interpreter/bytecodeTracer.cpp:99]: nit, can you change this variable name to method_changed for the C++ coding style.

**Review by coleenp — APPROVED** [review]
This seems fine.  I had the same question that David did about using UL, but this is simpler.  Most of the Trace* options in the runtime were converted to UL but where they have a specific format, it was too cumbersome so we didn't do it.

File: src/hotspot/share/interpreter/bytecodeTracer.cpp (hunk 1 above)
Comment (cl4es) [reply src/hotspot/share/interpreter/bytecodeTracer.cpp]: Done!

**Review by dholmes-ora — APPROVED** [review]

**Comment by cl4es** [comment]
Thanks for reviewing, everyone!

/integrate

**Pull request closed without merging**